### ISSUE TITLE
remote cache does not work with "compiler plugin" + "different coursier home(different OS or user.home)"

### ISSUE THREAD

**Issue by xuwei-k** [sbt/sbt#6027]
## steps

https://github.com/xuwei-k/sbt-remote-cache-compiler-plugin/commit/d4b466ccacb43ddf810879706775ba75a716c9ab

### A.scala

```scala
package example

class A
```

### `build.sbt`

```scala
name := "my-project"

scalaVersion := "2.13.3"

pushRemoteCacheTo := Some(
  MavenCache("local-cache", (ThisBuild / baseDirectory).value / "my_remote_cache")
)

remoteCacheId := "fixed-id"

remoteCacheIdCandidates := Seq(remoteCacheId.value)

pushRemoteCacheConfiguration := pushRemoteCacheConfiguration.value.withOverwrite(true)

addCompilerPlugin("org.typelevel" % "kind-projector" % "0.11.0" cross CrossVersion.full)
```

### `project/build.properties`

```properties
sbt.version=1.4.1
```

run:

```
sbt pushRemoteCache
rm -rf target
sbt -Dsbt.coursier.home=my_coursier_home pullRemoteCache compile
```

#### default coursier home dir

https://get-coursier.io/docs/cache.html#location

> - on Linux, ~/.cache/coursier/v1. This also applies to Linux-based CI environments, and FreeBSD too.
> - on OS X, ~/Library/Caches/Coursier/v1.
> - on Windows, %LOCALAPPDATA%\Coursier\Cache\v1, which, for user Alex, typically corresponds to C:\Users\Alex\AppData\Local\Coursier\Cache\v1.

https://github.com/xuwei-k/sbt-remote-cache-compiler-plugin/runs/1307158439#step:4:218

### build log

remote cache does not works. recompile 😢 

```
+ sbt -Dsbt.coursier.home=my_coursier_home pullRemoteCache compile
[info] welcome to sbt 1.4.1 (AdoptOpenJDK Java 1.8.0_272)
[info] loading project definition from /home/runner/work/sbt-remote-cache-compiler-plugin/sbt-remote-cache-compiler-plugin/project
[info] loading settings for project sbt-remote-cache-compiler-plugin from build.sbt ...
[info] set current project to my-project (in build file:/home/runner/work/sbt-remote-cache-compiler-plugin/sbt-remote-cache-compiler-plugin/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] remote cache artifact extracted for Some(cached-compile)
[info] remote cache artifact extracted for Some(cached-test)
[success] Total time: 1 s, completed Oct 26, 2020 5:56:57 AM
[info] compiling 1 Scala source to /home/runner/work/sbt-remote-cache-compiler-plugin/sbt-remote-cache-compiler-plugin/target/scala-2.13/classes ...
[info] done compiling
[success] Total time: 5 s, completed Oct 26, 2020 5:57:02 AM
```

Could not use retrieveManaged 😢  https://github.com/sbt/sbt/issues/5078

## expectation

don't recompile

## notes

https://github.com/sbt/sbt/blob/7a7bc5d0f933aa12988f81d2e18c3cfb538db129/main/src/main/scala/sbt/Defaults.scala#L3848

sbt use absolute path for `-Xplugin:` args.

**Comment by xuwei-k** [comment]
## workaround

https://github.com/xuwei-k/sbt-remote-cache-compiler-plugin/commit/b9fd9b6a66b8204e29e97d59895e31c6b38065b5
https://github.com/xuwei-k/sbt-remote-cache-compiler-plugin/runs/1307448414

```scala
Seq(Compile, Test).map { c =>
  c / scalacOptions := {
    val base = (LocalRootProject / baseDirectory).value
    val compilerPluginsTempDir = (LocalRootProject / target).value / "compiler_plugins"
    val prefix = "-Xplugin:"
    (c / scalacOptions).value.map { opt =>
      if (opt startsWith prefix) {
        val originalPluginFile = file(opt.drop(prefix.size))
        val pluginJarName = originalPluginFile.getName
        val targetJar = compilerPluginsTempDir / pluginJarName
        if(!targetJar.isFile) {
          IO.copyFile(
            sourceFile = originalPluginFile,
            targetFile = targetJar
          )
        }
        prefix + IO.relativize(base, targetJar).get
      } else {
        opt
      }
    }
  }
}
```

**Comment by xuwei-k** [comment]
https://github.com/sbt/sbt/blob/7a7bc5d0f933aa12988f81d2e18c3cfb538db129/main/src/main/scala/sbt/plugins/SemanticdbPlugin.scala#L82

I think `-P:semanticdb:targetroot:` is also use absolute path.

**Comment by eed3si9n** [comment]
That makes sense. We should pass `VirtualFileRef`, and maybe Zinc can change it to `Path` at the last minute.

**Comment by xuwei-k** [comment]
### workaround for SemanticdbPlugin

```scala
Seq(Compile, Test).flatMap { c =>
  (c / semanticdbTargetRoot) := {
    val old = (c / semanticdbTargetRoot).value.toPath
    (LocalRootProject / baseDirectory).value.toPath.relativize(old).toFile
  }
}
```

**Comment by pavelfatin** [comment]
Related: https://youtrack.jetbrains.com/issue/SCL-18534#focus=Comments-27-4614571.0-0

**Comment by kubukoz** [comment]
Just wanted to report that Kenji's workaround works really well for copy-pasting sources (which we use on CI), but nowadays it seems to break Semanticdb for Metals. :(

**Comment by adriaanm** [comment]
FYI, this relativizing to `(LocalRootProject / baseDirectory)` approach breaks workflows where you use something like [sbt/sbt-sriracha](https://github.com/sbt/sbt-sriracha) to build the equivalent of a monorepo using lots of `ProjectRef`s to link individual repos into one big build. Kind of like a community build using only sbt. (I'm enjoying building this at work.) E.g. a structure like:

```
sbt:buildall> show baseDirectory

[info] ProjectRef(uri("file://~/g/repo1/"), "someProject1") / baseDirectory
[info]  ~/g/repo1/someProject1Dir
....
[info] ProjectRef(uri("file://~/g/repoN/"), "someProjectM") / baseDirectory
[info]  ~/g/repoN/someProjectMDir
[info] baseDirectory
[info]  ~/g/buildall
```

When you then compile all these things together, you get lots of paths relative to each individual `~/g/repo$i/someProject$jDir/`, while they should be relativized to the root of your monorepo build, `~/g/buildall`.

**Comment by eed3si9n** [comment]
That's interesting.

So it seems like we need to satisfy 3 goals:
1. Convince Scala compiler to run a compiler plugin with JAR output and optionally other outputs like semanticdb
2. Make the outputs usable from across subprojects in a [source-dependency](https://eed3si9n.com/hot-source-dependencies-using-sbt-sriracha/) monorepo
3. Make the outputs available to IDEs

I feel like within each repo, using relative path against `LocalRootProject` for things like `-Xplugin` and `semanticdbTargetRoot ` makes sense to me. If we need to advertise semanticdb locations across builds, maybe we should create a different key?

> When you then compile all these things together, you get lots of paths relative to each individual `~/g/repo$i/someProject$jDir/`, while they should be relativized to the root of your monorepo build, `~/g/buildall`.

How is this manifesting as an error right now? Metals can't find semanticdb?

### sidenote

Despite its reputation for monorepo, Bazel often has to deal with multi-repo situations to load the plugins, and it makes each repo (workspace) declare a nickname for like `com_google_protobuf` instead of using `uri(...)`. This allows different workspaces to universally reference `@com_google_protobuf//:any_proto` etc, which sort of works. (Bzlmod I think is an on-going effort to make this meta-dependencies centralized on GitHub). Maybe it would be cool if `ProjectRef` could do something similar so the physical location of where you got "repo1" doesn't leak out, and it would just say `ProjectRef("@repo1", "someProject1")`.

**Comment by adriaanm** [comment]
> How is this manifesting as an error right now? Metals can't find semanticdb?

yes, it varies between plugins, but, for example, with wartremover excludes don't work because the directory is not found. I haven't had time to dig in with the issues I'm seeing in other plugins (metals, sbt-protoc), but they look similar.

The only "easy" fix that I've come up with is to factor out the `(LocalRootProject / baseDirectory).value` directory that's used to hide the path from root to the workspace (as it varies between machines) into a key. It could default to the value of `sbt.workspace` from your [sriracha post](https://eed3si9n.com/hot-source-dependencies-using-sbt-sriracha/). (I think you're suggesting the same thing in your comment.)

Since this is a common pattern across plugins, maybe you could consider introducing an `sbtWorkspaceDirectory` key in sbt?

Another potential solution -- I don't know how feasible --  could be to have something like `GlobalRootProject`, which would point to the "true root" build that is depending on an external build via a `ProjectRef`. That way, I could set `baseDirectory` in my real root project, and all "imported" projects would use that.
